### PR TITLE
fix: Add missing default styling of `PhysicalComponent`s for natures

### PR DESCRIPTION
**Pull request by ewuerger**
Does solve #167 but not ideally and dry.

Have a look at the [capstyle default style dict](https://github.com/DSD-DBS/py-capellambse/compare/master...fix-phys-nature-styling#diff-2ce9560a92a80b69c73bea77215b9e2a0297f05c6b377e8edb5e054587e36cd3L747-R783). With this solution we needed to add many styleclasses with the same style-dict due to the `part_factory`. We dynamically add `Node` or `Behavior` to the `.styleclass` attribute of the object like we did in the `Human` and `Actor`  case. Indeed it seems like the `is_human` attribute (causing `Human` in the styleclass) doesn't change the styling in anyway. In the stylesheet generation it seems like we are matching the whole styleclass which limits us to also use the whole styleclass in the key in capstyle's default styledict. We have 2 ways to improve the situation:

1. Make the stylesheet generator compatible with wildcards/regex. This enables usage of styleclasses like e.g.: `Physical*Actor`, `Physical*NodeComponent` and `Physical*BehaviorComponent` and these 3 would be the only needed, differing style pairs.
2. Instead of having one styleclass add support of multiple classes, where the first one is the inferred/original one from the xtype (here `PhysicalComponent`). This would lead to styleclasses `PhysicalComponent`, `Human`, `Actor`, `Node` or `Behavior`. How would we determine which styling class is superior to the others? Take in mind this option does need more changes, especially in the file that declares diagram objects (`Diagram`, `Box`, `Edge`, etc.) and the svg submodule.

I'm looking forward to discuss moving forward.

### CODE REVIEW

**Comment by Wuestengecko** [comment]
> Make the stylesheet generator compatible with wildcards/regex.

Bad idea, this would give up all the performance benefits of hashable keys, and AIRD diagram rendering is already slow enough.

> Instead of having one styleclass add support of multiple classes, where the first one is the inferred/original one from the xtype (here PhysicalComponent).

That could work, but we'd still need to distinguish between different derived classes within one "original" class. For example, the `PhysicalBehaviorActor` and `PhysicalBehaviorComponent` introduced in this very PR both derive from `PhysicalComponent`. So just specifying the original class wouldn't be enough, we'd need to make up a separate "styling-only class" ... which would introduce a not-insignificant amount of complexity to this already very complex module.

---

Another option could be to have some kind of dynamism in the definition itself. Right now the definitions are static: they're looked up by a predefined name and that's it, if you want a different definition you need to change the name you look up. We could make it possible to define a function instead that takes a (semantic) model element to be styled, and returns the default styling for that element. Then we might not even need to modify the `styleclass` at all like we currently do.

But this would probably also have some ripple effects in the SVG part of the discussion (as it also defers to `capstyle`), and in general is a large enough change that I would postpone it to when we attack #107 - one more breaking change there wouldn't stand out ;)

**Review by Wuestengecko — CHANGES_REQUESTED** [review]

File: capellambse/aird/parser/_box_factories.py [hunk 1]
```diff
@@ -306,6 +306,27 @@ def part_factory(seb: C.SemanticElementBuilder) -> aird.Box:
     raise C.SkipObject()
 
 
+def physical_part_factory(seb: C.SemanticElementBuilder) -> aird.Box:
```
[the remark this answers is not in the document]
Comment (ewuerger) [reply capellambse/aird/parser/_box_factories.py:309]: I now realize that for physical `Part`s we'd need the styleclass-patch-logic implemented via the `physical_part_factory`, but for `Part`s the part box factory is called... I'll merge the logic into the `part_factory`.

File: capellambse/aird/capstyle.py [hunk 2]
```diff
@@ -744,14 +744,43 @@ class in the form::
...
+                    COLORS["_CAP_Node_Yellow"],
+                ],
+                "stroke": COLORS["_CAP_Node_Yellow_Border"],
+                "text_fill": COLORS["_CAP_Node_Yellow_Label"],
+            }
+        ),
+        "Box.PhysicalHumanBehaviorActor": PHYS_ACTOR,  # type: ignore[dict-item]
+        "Box.PhysicalHumanNodeActor": PHYS_ACTOR,  # type: ignore[dict-item]
+        "Box.PhysicalNodeComponent": PHYS_NODE,  # type: ignore[dict-item]
+        "Box.PhysicalNodeActor": PHYS_ACTOR,  # type: ignore[dict-item]
```
[the remark this answers is not in the document]
Comment (ewuerger) [reply capellambse/aird/capstyle.py:783]: How? Mypy gives me this:
> Dict entry 3 has incompatible type "str": "Dict[str, Sequence[object]]"; expected "str": "Dict[str, Union[int, str, RGB, Sequence[RGB], None]]"  [dict-item]mypy(error)

File: capellambse/aird/parser/_box_factories.py [hunk 3]
```diff
@@ -306,6 +306,27 @@ def part_factory(seb: C.SemanticElementBuilder) -> aird.Box:
...
+    """Resolve a physical ``Part`` meta-styleclass and create its Box."""
+    part = part_factory(seb)
+    data_obj = seb.melodyobjs[-1]
```
[the remark this answers is not in the document]
Comment (ewuerger) [reply capellambse/aird/parser/_box_factories.py]: I don't want to assume things. So I integrated the logic into the `part_factory` styleclass altering section.

**Comment by ewuerger** [comment]
>     Instead of having one styleclass add support of multiple classes, where the first one is the
>     inferred/original one from the xtype (here PhysicalComponent).
> 
> That could work, but we'd still need to distinguish between different derived classes within one "original" class. For example, the PhysicalBehaviorActor and PhysicalBehaviorComponent introduced in this very PR both derive from PhysicalComponent. So just specifying the original class wouldn't be enough, we'd need to make up a separate "styling-only class" ... which would introduce a not-insignificant amount of complexity to this already very complex module.

What I meant is that we'll always keep the original styleclass that is received from the xtype (e.g.: `PhysicalComponent`). This one has some style definitions in capstyle's style-dict. Now we add other, *custom* styleclasses (e.g.: `Actor` or `Node`) that come with their own default style-dicts, maintained in capstyle's style-dict. The problem then is which overlapping CSS-definitions from the custom default styles are superior and to be taken...

Also: Even **without** having the default styles registered in capstyle's style-dict, the correct colors and gradients (I suppose `obj_styles` but not `text_styles`!) are recognized in the rendering step. We may invest too much time into default style declaration that we never use... I'm not 100% sure about that.

File: capellambse/aird/parser/_box_factories.py [hunk 4]
```diff
@@ -290,10 +290,12 @@ def part_factory(seb: C.SemanticElementBuilder) -> aird.Box:
         seb.styleclass = abstract_obj.attrib[C.ATT_XST].split(":")[-1]
         assert seb.styleclass is not None
         if seb.styleclass.endswith("Component"):
+            nature = ("Behavior", "Node")[abstract_obj.get("nature") == "NODE"]
             seb.styleclass = "".join(
                 (
                     seb.styleclass[: -len("Component")],
                     "Human" * (abstract_obj.get("human") == "true"),
+                    nature * (bool(abstract_obj.get("nature"))),
```
Comment (Wuestengecko) [inline capellambse/aird/parser/_box_factories.py:298]: This looks very weird to me. You're setting `nature` to one of two values depending on an XML attribute, but then here you're double-checking whether that attribute was set in the first place? Just set the variable to the correct value from the start. Keeps the logic together and avoids having to double-check.

**Review by Wuestengecko — CHANGES_REQUESTED** [review]

**Comment by Wuestengecko** [comment]
> always keep the original styleclass that is received from the xtype […]. Now we add other, _custom_ styleclasses (e.g.: `Actor` or `Node`) that come with their own default style-dicts, maintained in capstyle's style-dict. The problem then is which overlapping CSS-definitions from the custom default styles are superior and to be taken...

That would be a problem that I don't really think we should have to deal with if we can avoid it. Although I guess dealing with it could also be as simple as making the ordering of those additional classes significant, and the last one wins. But still, IMHO having to deal with multiple style classes at all is already too much complexity.

> Also: Even **without** having the default styles registered in capstyle's style-dict, the correct colors and gradients (I suppose `obj_styles` but not `text_styles`!) are recognized in the rendering step. We may invest too much time into default style declaration that we never use... I'm not 100% sure about that.

That happens because Capella often saves all style information explicitly in the AIRD, and the parser picks it up if it differs from its known defaults. But our use case also entails supporting other diagram sources like the context diagrams extension, which greatly benefits from having accurate defaults. So IMO this is time well spent.

File: capellambse/aird/capstyle.py [hunk 5]
```diff
@@ -744,43 +769,14 @@ class in the form::
...
-        "Box.PhysicalNodeComponent": PHYS_NODE,  # type: ignore[dict-item]
-        "Box.PhysicalNodeActor": PHYS_ACTOR,  # type: ignore[dict-item]
+        "Box.PhysicalBehaviorActor": _PHYS_ACTOR,
+        "Box.PhysicalBehaviorComponent": _PHYS_BEHAVIOR,
+        "Box.PhysicalHumanBehaviorComponent": _PHYS_BEHAVIOR,
+        "Box.PhysicalHumanNodeComponent": _PHYS_NODE,
+        "Box.PhysicalHumanBehaviorActor": _PHYS_ACTOR,
+        "Box.PhysicalHumanNodeActor": _PHYS_ACTOR,
+        "Box.PhysicalNodeComponent": _PHYS_NODE,
+        "Box.PhysicalNodeActor": _PHYS_ACTOR,
```
[the remark this answers is not in the document]
Comment (ewuerger) [reply capellambse/aird/capstyle.py:779]: The different symbols/decorations are not yet implemented. This is tbd.

[the remark this answers is not in the document]
Comment (Wuestengecko) [reply capellambse/aird/capstyle.py:783]: "Not yet implemented" alright sure, and IMO adding the icons is out of scope of this PR. But we do need to figure out which icons we'll eventually need, in order to properly define the style classes. And unfortunately it does seem that there are actually differences in the icons between all 8 cases for icons.

So, the answer to my questions above is "we do need to distinguish between them for icons".

**Comment by ewuerger** [comment]
We need to do this in a better way. With this approach we break existing decorations on `PhysicalComponent`s. Thank god these are the only ones that have a `nature` attribute`.

**Comment by ewuerger** [comment]
I think we can close this stale PR which is still in draft. @Wuestengecko ?

**Comment by Wuestengecko** [comment]
Yeah, let's just close it. It has been rotting away for a while now, and at this point doing it from scratch would probably be easier than trying to rebase anyways.

**Pull request closed without merging**